### PR TITLE
implement a new API

### LINKED ISSUE #506: Re-enable skipped client test
As of #503, the "doesn't return after version negotiation is established if no ConnState is defined" test is disabled.

### LINKED ISSUE #563: handle handshake errors in h2quic.Client
If the crypto handshake fails, the `h2quic.Client.Do` should return immediately.
At the moment, it is stuck waiting at the `cryptoChangedCond` `sync.Cond`, and will never return.

### LINKED ISSUE #564: return an error in Dial() if the handshake fails
Related to #563.

### LINKED ISSUE #566: think about replacing the ConnStateCallback by a channel
This came up in the discussion of #564. I don't understand how it helps there, but it might be an idea worth exploring on its own.
At the moment, users of our API have to deal with the fact that the `ConnStateCallback` is called in a separate goroutine, i.e. a common practice will be to protect the `ConnStateCallback` by a mutex (see for example [client.go#L69](https://github.com/lucas-clemente/quic-go/blob/master/h2quic/client.go#L69) . Using a channel, this could probably be avoided.

### LINKED ISSUE #568: Rethink API
Current new API proposal after discussion with @marten-seemann 

```go
// Server

listener, err := quic.ListenAddr(addr, cfg)
go listener.Serve() // <-- problematic, could be in ListenAddr()

for {
	session, err := listener.Accept()
	go handleSession(session)
}

// For non-fw cases

type NonFwSession interface {
	Session
	WaitUntilForwardSecure() error
}

for {
	session, err := listener.AcceptNonForwardSecure()
	// session is of type NonFwSession, which is
	go func() {
		doNonFwWork(session)
		err := session.WaitUntilForwardSecure()
		doWork(session)
	}()
}

// Client

session, err := quic.DialAddr(addr, cfg)
// session is fw-secure here

// For non-fw cases

fwSecureSession, err := quic.DialAddr(addr, &quic.Config{
	OnNonForwardSecure: func(nonFwSecureSession Session) {
		/* go */ doNonFwWork(nonFwSecureSession)
	},
})
// we block, so fwSecureSession here is fw secure

// Client #2

session, err := quic.DialAddr(addr, cfg)
// session is fw-secure here

// For non-fw cases
session, err := quic.DialAddrNonForwardSecure(addr, cfg)
// session is non-fw secure here
session.WaitUntilForwardSecure()
// session is fw-secure here
```

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #568, fixes #564, fixes #563.
Closes #566, closes #506 (obsolete now).

### CODE REVIEW

File: h2quic/client.go [hunk 1]
```diff
@@ -169,16 +138,14 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 
 	hasBody := (req.Body != nil)
 
-	c.mutex.Lock()
-	for c.encryptionLevel != protocol.EncryptionForwardSecure {
-		c.cryptoChangedCond.Wait()
-	}
+	<-c.handshakeChan // wait until the handshake has completed
```
Comment (lucas-clemente) [inline h2quic/client.go]: We should probably have a way to unblock this if an error occurs between dialing and closing the chan.

File: session.go [hunk 2]
```diff
@@ -83,10 +83,11 @@ type session struct {
...
 	// this channel is passed to the CryptoSetup and receives the current encryption level
 	// it is closed as soon as the handshake is complete
-	aeadChanged        <-chan protocol.EncryptionLevel
-	handshakeComplete  bool
-	handshakeChan      chan struct{} // will be closed as soon as the handshake completes
-	handshakeErrorChan chan error
+	aeadChanged       <-chan protocol.EncryptionLevel
+	handshakeComplete bool
+	// will be closed as soon as the handshake completes, and receive any error that might occur until then
+	// it to block WaitUntilHandshakeComplete()
```
Comment (lucas-clemente) [inline session.go]: Grammar is weird here.

File: client.go [hunk 3]
```diff
@@ -81,7 +81,21 @@ func Dial(pconn net.PacketConn, remoteAddr net.Addr, host string, config *Config
...
+func Dial(pconn net.PacketConn, remoteAddr net.Addr, host string, config *Config) (Session, error) {
+	sess, err := DialNonFWSecure(pconn, remoteAddr, host, config)
+	if err != nil {
+		return nil, err
+	}
+	err = sess.WaitUntilHandshakeComplete()
+	if err != nil {
+		return nil, err
+	}
+	return sess, nil
```
Comment (lucas-clemente) [inline client.go]: Can be simplified to `return sess, sess.WaitUntilHandshakeComplete()`

File: session.go [hunk 4]
```diff
@@ -159,17 +175,19 @@ func newClientSession(conn connection, hostname string, v protocol.VersionNumber
 
 	aeadChanged := make(chan protocol.EncryptionLevel, 2)
 	s.aeadChanged = aeadChanged
+	handshakeChan := make(chan handshakeEvent, 1)
```
Comment (lucas-clemente) [inline session.go]: Nit: don't really need the var here, can assign to s.handshakeChan directly.

File: session.go [hunk 5]
```diff
@@ -287,6 +311,8 @@ runLoop:
 	// otherwise this chan will already be closed
 	if !s.handshakeComplete {
 		s.handshakeCompleteChan <- closeErr.err
+		s.handshakeChan <- handshakeEvent{err: closeErr.err}
+
```
Comment (lucas-clemente) [inline session.go]: Nit: remove empty line

File: server.go [hunk 6]
```diff
@@ -229,9 +229,21 @@ func (s *server) handlePacket(pconn net.PacketConn, remoteAddr net.Addr, packet
...
 			_ = session.run()
-
 			s.removeConnection(hdr.ConnectionID)
 		}()
+
+		go func() {
+			for {
+				ev := <-handshakeChan
+				if ev.err != nil {
+					return
```
Comment (lucas-clemente) [inline server.go]: Should we log the err here, or is that done someplace else?

File: session.go [hunk 7]
```diff
@@ -95,6 +89,7 @@ type session struct {
 	handshakeCompleteChan chan error
 	// handshakeChan receives handshake events and is closed as soon the handshake completes
 	// the receiving end of this channel is passed to the creator of the session
+	// it receives at most 3 handshake event: 2 when the encryption level changes, and one error
```
Comment (lucas-clemente) [inline session.go]: nit: s/event/events/

File: session.go (hunk 4 above)
Comment (marten-seemann) [inline session.go]: We do. `session.handshakeChan` is a directed channel.

File: client.go (hunk 3 above)
Comment (marten-seemann) [inline client.go]: Only if you're willing to hand out the `sess`, even if `WaitUntilHandshakeComplete()` fails. I'd prefer to stick to the pattern that all return values are the respective null values if `err != nil`.

File: server.go (hunk 6 above)
Comment (marten-seemann) [inline server.go]: This error should get already get logged by session, when it closes.

**Comment by marten-seemann** [comment]
@lucas-clemente: Done.

File: h2quic/client.go [hunk 8]
```diff
@@ -44,57 +46,34 @@ var _ h2quicClient = &Client{}
...
 }
 
 // Dial dials the connection
-func (c *Client) Dial() error {
-	_, err := quic.DialAddr(c.hostname, c.config)
-	return err
-}
+func (c *Client) Dial() (err error) {
+	defer func() {
+		c.mutex.Lock()
```
Comment (lucas-clemente) [inline h2quic/client.go]: I don't think you actually need the mutex here if you always set the error before closing the chan.

File: h2quic/client.go [hunk 9]
```diff
@@ -170,17 +149,23 @@ func (c *Client) Do(req *http.Request) (*http.Response, error) {
 
 	hasBody := (req.Body != nil)
 
-	c.mutex.Lock()
-	for c.encryptionLevel != protocol.EncryptionForwardSecure {
-		c.cryptoChangedCond.Wait()
+	// wait until the handshake is complete
+	<-c.handshakeChan
+	c.mutex.RLock()
```
Comment (lucas-clemente) [inline h2quic/client.go]: Same here, I think this Lock is unnecessary.

File: client.go (hunk 3 above)
Comment (lucas-clemente) [inline client.go]: I wouldn't say it's necessary to return a null value if err != nil, that's what the err is for – but I'm fine with keeping it as-is if you insist.

File: h2quic/client.go [hunk 10]
```diff
@@ -44,57 +46,34 @@ var _ h2quicClient = &Client{}
...
 		t:               t,
+		dialAddr:        quic.DialAddr,
 		hostname:        authorityAddr("https", hostname),
 		responses:       make(map[protocol.StreamID]chan *http.Response),
 		encryptionLevel: protocol.EncryptionUnencrypted,
+		config: &quic.Config{
+			TLSConfig:                     tlsConfig,
+			RequestConnectionIDTruncation: true,
+		},
+		handshakeChan: make(chan struct{}),
```
Comment (lucas-clemente) [inline h2quic/client.go]: I think the name collision with the handshakeChan in client.go is unfortunate, could you maybe rename this one here?

**Comment by lucas-clemente** [comment]
LGTM modulo my mutex & renaming comments.

Please also update quic-conn at the same time as merging this.

**Comment by marten-seemann** [comment]
Merged. I just noticed we forgot to update the CHANGELOG.
We should probably also have another look at our godoc.

**Pull request merged**